pkg/services/ngalert/notifier/config.go

```diff
@@ -0,0 +1,99 @@
+package notifier
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/grafana/alerting-api/pkg/api"
+	"github.com/grafana/grafana/pkg/infra/log"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+func PersistTemplates(cfg *api.PostableUserConfig, path string) ([]string, bool, error) {
+	if len(cfg.TemplateFiles) < 1 {
+		return nil, false, nil
+	}
+
+	var templatesChanged bool
+	pathSet := map[string]struct{}{}
+	for name, content := range cfg.TemplateFiles {
+		if name != filepath.Base(filepath.Clean(name)) {
+			return nil, false, fmt.Errorf("template file name '%s' is not valid", name)
+		}
+
+		err := os.MkdirAll(path, 0750)
+		if err != nil {
+			return nil, false, fmt.Errorf("unable to create template directory %q: %s", path, err)
+		}
+
+		file := filepath.Join(path, name)
+		pathSet[file] = struct{}{}
+
+		// Check if the template file already exists and if it has changed
+		// We can safeily ignore gosec here and we've previously checked the filename is clean
+		// nolint:gosec
+		if tmpl, err := ioutil.ReadFile(file); err == nil && string(tmpl) == content {
+			// Templates file is the same we have, no-op and continue.
+			continue
+		} else if err != nil && !os.IsNotExist(err) {
+			return nil, false, err
+		}
+
+		if err := ioutil.WriteFile(file, []byte(content), 0644); err != nil {
+			return nil, false, fmt.Errorf("unable to create Alertmanager template file %q: %s", file, err)
+		}
+		// nolint:gosec
+
+		templatesChanged = true
+	}
+
+	// Now that we have the list of _actual_ templates, let's remove the ones that we don't need.
+	existingFiles, err := ioutil.ReadDir(path)
+	if err != nil {
+		log.Error("unable to read directory for deleting alertmanager templates", "err", err, "path", path)
+	}
+	for _, existingFile := range existingFiles {
+		p := filepath.Join(path, existingFile.Name())
+		_, ok := pathSet[p]
+		if !ok {
+			templatesChanged = true
+			err := os.Remove(p)
+			if err != nil {
+				log.Error("unable to delete template", "err", err, "file", p)
+			}
+		}
+	}
```

Comment:
We are not using the returned `templatesChanged` boolean right now but I can see how this logic can cause issue: when we use it and skip template re-creation, then on restart with same disk, this logic is going to find that the files are same as config and say that templates didn't change, and hence we might skip creation of template tree. So the safest way to do it would be having the list of files loaded in the memory and compare against that, hence on restart we won't miss it.

Reply:
But I guess that is nothing to do with `PersistTemplates` itself, it is doing its job. We could take care in alertmanager to not skip templates when it's loading for the first time. So nothing to do for now, but something to be kept in mind.

Reply:
Discussed this offline, it's all about the usage of `PersistTemplates` return values. Decided to leave it as is for now and we are not yet using the variable.

In my eyes, this is meant to save persisting the file to disk over and over as we poll for new configuration.